### PR TITLE
mysql, expression: support sql_mode 'PAD_CHAR_TO_FULL_LENGTH'

### PR DESCRIPTION
**Pull request by spongedu**
Support sql_mode `PAD_CHAR_TO_FULL_LENGTH`.
BTW: expressions in Conditions seems to be pushed down to kv layer, so currently this flag don't work in conditions.

### CODE REVIEW

File: mysql/const_test.go [hunk 1]
```diff
@@ -191,3 +191,22 @@ func (s *testMySQLConstSuite) TestHighNotPrecedenceMode(c *C) {
...
+	tk.MustExec("insert into t1 values ('xy');")
+	tk.MustExec("set sql_mode='';")
+	r := tk.MustQuery(`SELECT a='xy        ', char_length(a) FROM t1;`)
+	r.Check(testkit.Rows("0 2"))
+	r = tk.MustQuery(`SELECT count(*) FROM t1 WHERE a='xy        ';`)
+	r.Check(testkit.Rows("0"))
+	tk.MustExec("set sql_mode='PAD_CHAR_TO_FULL_LENGTH';")
+	r = tk.MustQuery(`SELECT a='xy        ', char_length(a) FROM t1;`)
+	r.Check(testkit.Rows("1 10"))
+	//TODO: FIXME
```
Comment (zz-jason) [inline mysql/const_test.go]: how add a issue about it ?

**Comment by zz-jason** [comment]
LGTM

File: mysql/const_test.go (hunk 1 above)
Comment (spongedu) [reply mysql/const_test.go]: ok

**Comment by breeswish** [comment]
Could you also add tests for varchars?

**Comment by coocood** [comment]
@spongedu 
It only applies to `CHAR(n)` type,  `VARCHAR(n)` is not affected.

**Comment by spongedu** [comment]
@coocood  yep, only `CHAR(n)` type should be affected. It seems that I forget the type checking. I'll fix it

**Comment by spongedu** [comment]
@coocood Done. PTAL

**Comment by shenli** [comment]
@coocood PTAL

File: expression/column.go [hunk 2]
```diff
@@ -26,6 +26,7 @@ import (
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/types/json"
 	"github.com/pingcap/tidb/util/codec"
+	"strings"
```
Comment (coocood) [inline expression/column.go]: import order need fix.

File: expression/column.go [hunk 3]
```diff
@@ -238,12 +243,23 @@ func (col *Column) EvalString(row types.Row, sc *variable.StatementContext) (str
...
 		return res, err != nil, errors.Trace(err)
 	}
 	if row.IsNull(col.Index) {
 		return "", true, nil
 	}
-	return row.GetString(col.Index), false, nil
+	val := row.GetString(col.Index)
+	if sc.PadCharToFullLength {
+		valLen := len([]rune(val))
+		if valLen < col.RetType.Flen && sc.PadCharToFullLength {
```
Comment (coocood) [inline expression/column.go]: `sc.PadCharToFullLength` is already checked.

**Comment by coocood** [comment]
@spongedu 
Where is the varchar test?

File: expression/column.go (hunk 2 above)
Comment (spongedu) [reply expression/column.go]: ok

File: expression/column.go (hunk 3 above)
Comment (spongedu) [reply expression/column.go]: @coocood  I think `sc.PadCharToFullLength` should be checked for both branch of `if col.GetType().Hybrid()` condition

**Comment by spongedu** [comment]
@coocood  add tests for `varchar` type . PTAL

**Comment by coocood** [comment]
LGTM

File: expression/column.go [hunk 4]
```diff
@@ -238,12 +243,23 @@ func (col *Column) EvalString(row types.Row, sc *variable.StatementContext) (str
...
 		return res, err != nil, errors.Trace(err)
 	}
 	if row.IsNull(col.Index) {
 		return "", true, nil
 	}
-	return row.GetString(col.Index), false, nil
+	val := row.GetString(col.Index)
+	if sc.PadCharToFullLength && col.GetType().Tp == mysql.TypeString {
+		valLen := len([]rune(val))
+		if valLen < col.RetType.Flen && sc.PadCharToFullLength {
```
Comment (coocood) [inline expression/column.go]: `sc.PadCharToFullLength` is duplicated with line 256

**Comment by spongedu** [comment]
Done @coocood

**Comment by coocood** [comment]
@zz-jason 
PTAL

**Pull request merged**